File: leetcode/leetcode_p1392/code_1392.py
# 力扣1392题，最长快乐前缀
# 【快乐前缀】是在原字符串中既是非空前缀也是后缀（不包括原字符串自身）的字符串。
# 给定一个字符串s, 返回它的最长快乐前缀，如果不存在，则返回一个空字符串


def find_happy_01(s):
    length = len(s)

    max_hp = 'non str'
    sub_s = ''
    for i,v in enumerate(s):
        sub_s += v
        # print('sub_s:', sub_s)
        if sub_s[0] == s[length-i-1]:
            print(sub_s[0], '==', s[length-i-1], sub_s)
            if sub_s == s[length-i-1:] and len(sub_s) != length:
                max_hp = sub_s

    print(max_hp)
    return max_hp


def find_happy_02(s):
    n = len(s)
    prefix, suffix = 0, 0
    base, mod, mul = 31, 1000000007, 1
    happy = 0
    for i in range(1, n):
        prefix = (prefix * base + (ord(s[i - 1]) - 97)) % mod
        suffix = (suffix + (ord(s[n - i]) - 97) * mul) % mod
        if prefix == suffix:
            happy = i
        mul = mul * base % mod
    return s[:happy]


def find_happy_03(s):
    n = len(s)
    record_list = [-1 for _ in range(n)]
    for i in range(1, n):
        j = record_list[i-1]
        while j != -1 and s[i] != s[j+1]:
            j = record_list[j]
        if s[j+1] == s[i]:
            record_list[i] = j + 1
    print(record_list)
    return s[:record_list[-1]+1]


s = 'acccbaaacccbaac'
# s = 'ababab'
result = find_happy_03(s)
print('result:', result)

File: leetcode/leetcode_p1392/code_1392.cpp
// 力扣1392题，最长快乐前缀
// 【快乐前缀】是在原字符串中既是非空前缀也是后缀（不包括原字符串自身）的字符串
// 给定一个字符串s，返回它的最长快乐前缀

#include <iostream>
#include <string>
#include <vector>
using namespace std;


class Solution_00{
public:
    string longestPrefix(string s){
        int n = s.size();
        string happy = "";
        for (int i = 1; i < n; ++i)
        {
            string prefix = s.substr(0, i);
            string suffix = s.substr(n-i, i);
            cout << prefix << " | " << suffix << endl;
            if (prefix == suffix)
                happy = prefix;
        }
        return happy;
    };
};

// 方法一：Rabin-Karp字符串编码
class Solution_01{
public:
    string longestPrefix(string s){
        int n = s.size();
        int prefix = 0, suffix = 0;
        int base = 31, mod = 1000000007, mul = 1;
        int happy = 0;
        for (int i = 1; i < n; ++i){
            prefix = ((long long)prefix * base + (s[i - 1] - 97)) % mod;
            suffix = (suffix + (long long)(s[n-i] - 97) * mul) % mod;
            if (prefix == suffix)
                happy = i;
            mul = (long long)mul * base % mod;
        }
        return s.substr(0, happy);
    };
};


// KMP算法
class Solution_02{
public:
    string longestPrefix(string s){
        int n = s.size();
        vector<int> fail(n, -1);
        for (int i = 1; i < n; ++i)
        {
            int j = fail[i-1];
            while (j != -1 && s[j+1] != s[i])
                j = fail[j];
            if (s[j+1] == s[i])
                fail[i] = j + 1;
        }
        return s.substr(0, fail[n-1]+1);
    }
};



int main()
{
    string s = "acccbaaacccbaac";
    cout << "s: " << s << endl;
    Solution_00 sl00;
    Solution_01 sl01;
    Solution_02 sl02;

    string s00 = sl00.longestPrefix(s);
    cout << "s00: " << s00 << endl;

    string s01 = sl01.longestPrefix(s);
    cout << "s01: " << s01 << endl;

    string s02 = sl02.longestPrefix(s);
    cout << "s02: " << s02 << endl;

    cin.get();

    return 0;
}